crates/sui/src/client_commands.rs

```diff
@@ -93,26 +94,116 @@ use crate::key_identity::{get_identity_address, KeyIdentity};
 #[cfg(test)]
 mod profiler_tests;
 
+/// Dry run, execute, or serialize a transaction.
+///
+/// This basically extracts the logical code for each command that deals with dry run, executing,
+/// or serializing a transaction and puts it in a single macro function to reduce code duplication.
+// TODO (stefan): Add gas_price option for all commands and remove it from this macro
 #[macro_export]
-macro_rules! serialize_or_execute {
-    ($tx_data:expr, $serialize_unsigned:expr, $serialize_signed:expr, $context:expr, $result_variant:ident) => {{
+macro_rules! dry_run_or_execute_or_serialize {
+    ($signer:expr, $tx_kind:expr, $context:expr, $gas_payment:expr, $gas_price:expr, $opts_with_gas:expr, $opts:expr, $result_variant:ident) => {{
+        let (
+            dry_run,
+            gas,
+            gas_budget,
+            serialize_unsigned_transaction,
+            serialize_signed_transaction,
+        ) = {
+            match $opts_with_gas {
+                Some(opts) => {
+                    (
+                        opts.rest.dry_run,
+                        opts.gas,
+                        opts.rest.gas_budget,
+                        opts.rest.serialize_unsigned_transaction,
+                        opts.rest.serialize_signed_transaction,
+                    )
+                },
+                None => {
+                    match $opts {
+                        Some(opts) => {
+                            (
+                                opts.dry_run,
+                                None,
+                                opts.gas_budget,
+                                opts.serialize_unsigned_transaction,
+                                opts.serialize_signed_transaction,
+                            )
+                        },
+                        None => (false, None, None, false, false)
+                    }
+                }
+            }
+        };
         assert!(
-            !$serialize_unsigned || !$serialize_signed,
+            !serialize_unsigned_transaction || !serialize_signed_transaction,
             "Cannot specify both --serialize-unsigned-transaction and --serialize-signed-transaction"
         );
-        if $serialize_unsigned {
-            SuiClientCommandResult::SerializedUnsignedTransaction($tx_data)
+        let gas_price = if let Some(gas_price) = $gas_price {
+            gas_price
+        } else {
+            $context.get_reference_gas_price().await?
+        };
+
+        let gas = match $gas_payment {
+            Some(obj_ids) => Some(obj_ids),
+            None => gas.map(|x| vec![x])
+        };
+
+        if dry_run {
+            execute_dry_run(
+                $context,
```

Comment:
When defining a macro, if you use the macro variable in multiple places, the expression it is bound to is going to be evaluated in multiple places, which is generally not what you want. It's better to create a binding for the result of that expression and use that in multiple places instead:

```rust
let context = $context;
```

But in the case of this macro (and the `serialize_or_execute` macro that existed before it) I can't actually tell why it needs to be a macro, rather than a function. The only macro-ish feature it's using is the `$result_variant: ident` parameter, but this is used at the very end to wrap a `response` variable which is always of the same type, so I think you might as well create a function instead -- it would avoid these kinds of complications.

Reply:
Good call! Refactored 🦤